File: Project/StudentMangerment/BusinessLogicLayer/Services/UserService.cs
using BusinessLogicLayer.Services.Interface;
using DataAccessLayer.Models;
using DataAccessLayer.Repositories.Interface;
using Microsoft.EntityFrameworkCore;

namespace BusinessLogicLayer.Services
{
    public class UserService : IUserService
    {
        private readonly IUserRepository _userRepo;

        public UserService(IUserRepository userRepo)
        {
            _userRepo = userRepo;
        }

        public async Task<(IEnumerable<User> Users, int Total)> GetPagedUsersAsync(
         string search, int pageIndex, int pageSize, int? roleId, bool? status)
        {
            var query = _userRepo.GetAllQueryable();

            // Search 
            if (!string.IsNullOrWhiteSpace(search))
            {
                search = search.Trim();
                query = query.Where(u =>
                    u.Username.Contains(search) ||
                    u.FullName.Contains(search) ||
                    u.Email.Contains(search) ||
                    u.Phone.Contains(search)
                );
            }

            // Filter role
            if (roleId.HasValue)
            {
                query = query.Where(u => u.UserRoles.Any(ur => ur.RoleId == roleId.Value));
            }

            // Filter status
            if (status.HasValue)
            {
                query = query.Where(u => u.IsStatus == status.Value);
            }

            // Count before paging
            int total = await query.CountAsync();

            // Get data with paging
            var users = await query
                .Include(u => u.UserRoles)
                    .ThenInclude(ur => ur.Role)
                .OrderBy(u => u.Username)
                .Skip((pageIndex - 1) * pageSize)
                .Take(pageSize)
                .ToListAsync();

            return (users, total);
        }


        public async Task<User> GetByIdAsync(int id) => await _userRepo.GetByIdAsync(id);

        public async Task AddAsync(User user) => await _userRepo.AddAsync(user);

        public async Task UpdateAsync(User user) => await _userRepo.UpdateAsync(user);

        public async Task DeleteAsync(int id) => await _userRepo.SoftDeleteAsync(id);
    }

}

File: VBNet/MyLibrary.BLL/Services/Implementations/UserService.vb
Imports System.Data.Entity
Imports MyLibrary.DAL
Imports MyLibrary.Domain.MyApp.Domain.Enums

Public Class UserService
    Implements IUserService

    Private ReadOnly _uow As UnitOfWork

    Public Sub New(uow As UnitOfWork)
        _uow = uow
    End Sub

    Public Async Function GetPagedAsync(keyword As String, status As Boolean?, pageIndex As Integer,
                             pageSize As Integer) As Task(Of PagedResult(Of UserDto)) Implements IUserService.GetPagedAsync


        Dim query = _uow.Users.GetAllIncludedDeleted().Include("UserRoles.Role")

        query = query.Where(Function(u) Not u.UserRoles.Any(Function(ur) ur.Role.RoleName = "Admin"))

        If Not String.IsNullOrWhiteSpace(keyword) Then
            Dim kw = keyword.Trim().ToLower()
            query = query.Where(Function(u) u.FullName.ToLower().Contains(kw) OrElse
                                            u.Email.ToLower().Contains(kw) OrElse
                                            u.Phone.Contains(kw))
        End If


        If status.HasValue Then
            query = query.Where(Function(u) u.IsActive = status.Value)
        End If

        Dim totalRow = Await query.CountAsync()

        Dim items = query.OrderByDescending(Function(u) u.Id) _
                         .Skip((pageIndex - 1) * pageSize) _
                         .Take(pageSize) _
                         .ToList() _
                         .Select(Function(u) New UserDto With {
                             .Id = u.Id,
                             .FullName = u.FullName,
                             .Email = u.Email,
                             .Phone = u.Phone,
                             .IsActive = u.IsActive,
                             .RoleNames = String.Join(", ", u.UserRoles.Select(Function(ur) ur.Role.RoleName))
                         }).ToList()

        Return New PagedResult(Of UserDto) With {
            .Items = items,
            .TotalCount = totalRow,
            .TotalPages = Math.Ceiling(totalRow / pageSize)
        }
    End Function

    Public Async Function ToggleStatusAsync(userId As Integer) As Task Implements IUserService.ToggleStatusAsync
        Dim user = Await _uow.Users.GetByIdAsync(userId)
        If user Is Nothing Then Throw New Exception("Người dùng không tồn tại")

        user.IsActive = Not user.IsActive

        _uow.Users.Update(user)
        Await _uow.SaveAsync()
    End Function
End Class